In [1]:
# Install all the packages
!pip install numpy
!pip install pandas
!pip install torch

In [3]:
# Import all the packages 
import numpy as np
import pandas as pd
from sklearn import preprocessing
from sklearn.model_selection import train_test_split  
import torch
import torch.nn as nn
import torch.optim as optim

url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"

In [5]:
# Let's start by naming the features
names = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'Class']

In [7]:
# Reading the dataset through pandas function
iris_data = pd.read_csv( url , names = names )
print(iris_data)

     sepal-length  sepal-width  petal-length  petal-width           Class
0             5.1          3.5           1.4          0.2     Iris-setosa
1             4.9          3.0           1.4          0.2     Iris-setosa
2             4.7          3.2           1.3          0.2     Iris-setosa
3             4.6          3.1           1.5          0.2     Iris-setosa
4             5.0          3.6           1.4          0.2     Iris-setosa
..            ...          ...           ...          ...             ...
145           6.7          3.0           5.2          2.3  Iris-virginica
146           6.3          2.5           5.0          1.9  Iris-virginica
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

[150 rows x 5 columns]


In [9]:
# Analyzing the table 
iris_data_analyze = iris_data.info()
print(iris_data_analyze)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal-length  150 non-null    float64
 1   sepal-width   150 non-null    float64
 2   petal-length  150 non-null    float64
 3   petal-width   150 non-null    float64
 4   Class         150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB
None


In [11]:
# Checking for missing values 
missing_values = iris_data.isnull().sum()
print(missing_values)

sepal-length    0
sepal-width     0
petal-length    0
petal-width     0
Class           0
dtype: int64


In [13]:
# Assigning the inputs with feature variables
inputs = iris_data.drop(columns = ['Class'])
# Assigning the outputs with targret variables 
outputs = iris_data['Class']

In [15]:
# Displaying the top section of inputs 
inputs.head()

sepal-length  sepal-width  petal-length  petal-width
0           5.1          3.5           1.4          0.2
1           4.9          3.0           1.4          0.2
2           4.7          3.2           1.3          0.2
3           4.6          3.1           1.5          0.2
4           5.0          3.6           1.4          0.2

In [17]:
# Displaying the top section of outputs 
outputs.head()

0    Iris-setosa
1    Iris-setosa
2    Iris-setosa
3    Iris-setosa
4    Iris-setosa
Name: Class, dtype: object

In [19]:
# Converting into one dimensinol numpy array using numpy 
outputs_1d = outputs.values

In [21]:
# Converting non-numerical values into numerical values using LabelEncoder
# Outputs(target variables) data structure is object type
le = preprocessing.LabelEncoder()
outputs_encoded = le.fit_transform(outputs_1d)
# Displaying the converted values 
print(outputs_encoded)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


In [23]:
# Spliting dataset for training(80%) and testing(20%)
inputs_train, inputs_test, outputs_encoded_train, outputs_encoded_test = train_test_split(inputs, outputs_encoded, test_size = 0.20)


In [25]:
# Normalizing the inputs data for both training and testing 
# Target variables are not normalized
# For Normalization we are using standardization(Z-score normalization) 
# Initialing the scaler
scaler = preprocessing.StandardScaler()
# Scalering the inputs data (features variables) of training 
inputs_train_normalized = scaler.fit_transform(inputs_train)
# Scalering the inputs data (features variables) of testing 
inputs_test_normalized = scaler.fit_transform(inputs_test)
# Displaying the size of Normalized data for training
print(inputs_train_normalized.size)
# Displaying the size of Normalized data for testing
print(inputs_test_normalized.size)

480
120


In feature variables there are 150 rows and 4 columns. So the total number of inputs data are 600(150*4). Therefore, training inputs = 480 (600 * 0.8) and testing inputs = 120 (600 * 0.2). Target variables are not normalized. After normalization, the data structure is always converted to numpy array. So, while using pytorch we need to convert the data structure in pytorch tensor. 

In [28]:
 # Convert the dataset to PyTorch tensors
c_inputs_train_normalized = torch.tensor(inputs_train_normalized, dtype=torch.float32)
c_inputs_test_normalized = torch.tensor(inputs_test_normalized, dtype=torch.float32)
c_outputs_encoded_train = torch.tensor(outputs_encoded_train, dtype=torch.long)
c_outputs_encoded_test = torch.tensor(outputs_encoded_test, dtype=torch.long)

In [30]:
# Designing MLP model
# Constructor : initializing 
class MLP(nn.Module):
    def __init__(self, input_size, hidden_layer1, hidden_layer2, hidden_layer3, output_size):
        super(MLP, self).__init__()
    
        # Designing the layer  
        self.input_hidden1_layer = nn.Linear(input_size, hidden_layer1)
        self.hidden1_hidden2_layer = nn.Linear(hidden_layer1, hidden_layer2)
        self.hidden2_hidden3_layer = nn.Linear(hidden_layer2, hidden_layer3)
        self.hidden3_output_layer = nn.Linear(hidden_layer3, output_size)
        # Defining the Activation Function: using rectified linear unit
        self.activation_function = nn.ReLU()

    # Passing the inputs into forward direction and computing
    def feedforwarding(self, inputs):
        computed_first_conn = self.input_hidden1_layer(inputs)
        computed_first_conn_with_activation = self.activation_function(computed_first_conn)
        computed_second_conn = self.hidden1_hidden2_layer(computed_first_conn_with_activation)
        computed_second_conn_with_activation = self.activation_function(computed_second_conn)
        computed_third_conn = self.hidden2_hidden3_layer(computed_second_conn_with_activation)
        computed_third_conn_with_activation = self.activation_function(computed_third_conn)
        computed_outputs = self.hidden3_output_layer(computed_third_conn_with_activation)
        return computed_outputs

In [32]:
# For this model, initialing
# inputs size will number of features
# Three hidden layer with 14,7,3 nodes
# output_size wii be one
# For evaluation purpose with other model this values are used
input_size = c_inputs_train_normalized.shape[1] # counting the number of features
hidden_layer1 = 14
hidden_layer2 = 7
hidden_layer3 = 3
output_size = 3
model = MLP(input_size, hidden_layer1, hidden_layer2, hidden_layer3, output_size)

In [34]:
# Defining the loss function 
criterion = nn.CrossEntropyLoss()
# Defining the optimizer: LBFGS Ooptimizer is used
optimizer = optim.LBFGS(model.parameters(), lr= 0.01)

In [36]:
# Training the model
def closure():
    # set to training mode
    model.train()
    # zero the gradients
    optimizer.zero_grad()
    # Forward feeding 
    outputs = model.feedforwarding(c_inputs_train_normalized)
    # calculating the loss
    loss = criterion(outputs, c_outputs_encoded_train)
    # Backward chaining 
    loss.backward()
    return loss



# Perform LBFGS optimization step: 
# LBFGS optimizer has its own looping rules
# So we don't need to perform the iteration
loss = optimizer.step(closure)
print(f'loss : {loss}')
  

loss : 1.1000912189483643


In [38]:
# Testing the model with test dataset
# set to testing mode
model.eval()
with torch.no_grad():
    outputs = model.feedforwarding(c_inputs_test_normalized)
    _, predicted = torch.max(outputs,1)
    accuracy = (predicted == c_outputs_encoded_test).sum().item() / c_outputs_encoded_test.size(0)
    print(f'Accuracy on test data: {accuracy:.4f}')


Accuracy on test data: 0.2333
